File: src/js/src/mapping/equal-interval-breaks.ts
import {initWASM} from '../init';
import {vecDoubleToNumber} from '../utils';

/**
 * ## Description
 * Divides the range of values into equal-sized intervals.
 * 
 * ## Characteristics
 * - Simple to understand and interpret
 * - Best for evenly distributed data
 * - May not represent data well when distribution is skewed
 * - Interval size = (maximum value - minimum value) / number of classes
 * 
 * @example
 * ```ts
 * import { equalIntervalBreaks } from 'geoda-wasm';
 * 
 * const data = [1, 2, 3, 4, 5, 6, 7, 8, 9];
 * const k = 3;
 * const breaks = await equalIntervalBreaks(k, data);
 * 
 * // breaks = [3.66666666666667, 6.33333333333333]
 * ```
 * 
 * @param k The number of classes/categories
 * @param data The numeric values to be classified
 * @returns The breaks values
 */
export async function equalIntervalBreaks(
  k: number,
  data: number[] | Float32Array
): Promise<number[]> {
  const wasm = await initWASM();

  const n = data.length;

  const wasmUndefs = new wasm.VectorInt();
  wasmUndefs.resize(n, 0);

  const wasmData = new wasm.VectorDouble();
  wasmData.resize(n, 0);
  for (let i = 0; i < n; ++i) {
    wasmData.set(i, data[i]);
    if (data[i] === undefined || data[i] === null) {
      wasmUndefs.set(i, 1);
    }
  }

  const result = wasm.equalIntervalBreaks(k, wasmData, wasmUndefs);

  return vecDoubleToNumber(result);
}

File: cpp/src/mapping/equal-interval-breaks.cpp
// SPDX-License-Identifier: MIT
// Copyright contributors to the geoda-lib project

#include <algorithm>
#include <limits>
#include <vector>

#include "mapping/mapping.h"

/**
 * @brief Calculates break points using the equal interval classification method
 *
 * This function divides the range of values into k equal-sized intervals.
 * It ignores undefined values when calculating the min and max values.
 *
 * @param k The number of classes/intervals desired
 * @param data Vector of numerical values to be classified
 * @param undefs Vector indicating which values in data are undefined (1=undefined, 0=defined)
 * @return std::vector<double> A vector of k-1 break points that define k intervals
 *
 * @note The returned breaks array will have size k-1, as k intervals are defined by k-1 break points
 * @note Values marked as undefined in undefs vector are excluded from min/max calculations
 */
std::vector<double> geoda::equal_interval_breaks(int k, const std::vector<double>& data,
                                                 const std::vector<int>& undefs) {
  double min_val = std::numeric_limits<double>::max();
  double max_val = std::numeric_limits<double>::lowest();
  for (int i = 0; i < data.size(); ++i) {
    if (undefs[i] == 1) continue;
    min_val = std::min(min_val, data[i]);
    max_val = std::max(max_val, data[i]);
  }
  double interval = (max_val - min_val) / k;
  std::vector<double> breaks(k - 1);
  for (int i = 0; i < k - 1; ++i) {
    breaks[i] = min_val + (i + 1) * interval;
  }
  return breaks;
}
